utility/include/isobus/utility/thread_synchronization.hpp

```diff
@@ -88,16 +98,368 @@ class LockFreeQueue
 	std::queue<T> queue; ///< The queue
 };
 
+#elif defined USE_CMSIS_RTOS2_THREADING
+
+#include "cmsis_os.h"
+
+#include <atomic>
+#include <cstdint>
+#include <functional>
+#include <vector>
+
+namespace isobus
+{
+	/// @brief A wrapper around a CMSIS RTOS 2 mutex.
+	/// @details See definition at https://www.keil.com/pack/doc/CMSIS/RTOS2/html/group__CMSIS__RTOS.html
+	class Mutex
+	{
+	public:
+		/// @brief Constructor for the CMSIS RTOS2 mutex wrapper
+		Mutex() :
+		  handle(nullptr)
+		{
+		}
+
+		/// @brief Locks the mutex. Part of BasicLockable requirements.
+		void lock()
+		{
+			if (ready())
+			{
+				osStatus_t osRetVal = osMutexAcquire(handle, osWaitForever);
+
+				if (osOK != osRetVal)
+				{
+					while (true)
+					{
+						// If your code is stuck in here, that means you did something
+						// very wrong, like recursively locked this mutex, or tried to
+						// lock the mutex before the OS was initialized, or called this in
+						// an interrupt service routine.
+						// osRetVal may contain more information.
+					}
```

Comment:
I like the way you used these for loops. It's a clever trick to help with finding where things go wrong with debugging. Though I wonder if it's always the way to go in production? I'm not sure